From 9c3520e6d5fcde225d55dc6a593d3c421954ccdb Mon Sep 17 00:00:00 2001
From: Ganesh Kariganuru Mahabalesh <quic_gkarigan@quicinc.com>
Date: Wed, 10 Apr 2024 00:31:36 -0700
Subject: [PATCH] hostapd: set NULL for driver private data for partner links 
 after deinitializing driver interface

In wifi down scenario, we are honoring NL80211_CMD_STOP_AP from
kernel which frees up links(decrements n_links value) using
nl80211_remove_link() and during the normal deinit path, we will check
n_links value to deinit nl80211 driver interface. Since we are cleaning up
links in multiple paths, while deinitializing the driver interface we will
encounter a double free scenario.

Hence, set NULL for driver private data for all the partner links after
deinit of the driver interface and refactor wpa_driver_get_shared_status
function.

Signed-off-by: Ganesh Kariganuru Mahabalesh <quic_gkarigan@quicinc.com>

--- a/src/ap/hostapd.c
+++ b/src/ap/hostapd.c
@@ -3490,7 +3490,7 @@ static void hostapd_cleanup_driver(const
 #endif /* CONFIG_IEEE80211BE */
 
 	if (driver->is_drv_shared &&
-	    !driver->is_drv_shared(drv_priv, iface->bss[0])) {
+	    !driver->is_drv_shared(drv_priv, iface->bss[0]->mld_link_id)) {
 		driver->hapd_deinit(drv_priv);
 		hostapd_bss_free(iface->bss[0]);
 		goto exit;
@@ -5287,7 +5287,7 @@ void hostapd_bss_free(struct hostapd_dat
 {
 	struct hostapd_data *link_bss = NULL;
 
-	if (!hapd)
+	if (!hapd || !hapd->conf->mld_ap)
 		return;
 
 	hapd->drv_priv = NULL;
--- a/src/drivers/driver_nl80211.c
+++ b/src/drivers/driver_nl80211.c
@@ -11134,11 +11134,12 @@ static int driver_nl80211_link_remove(vo
 }
 
 
-static bool nl80211_is_drv_shared(void *priv, void *bss_ctx)
+static bool nl80211_is_drv_shared(void *priv, int link_id)
 {
 	struct i802_bss *bss = priv;
 	struct wpa_driver_nl80211_data *drv = bss->drv;
 	unsigned int num_bss = 0;
+	int i;
 
 	/* If any other BSS exist, someone else is using this since at this
 	 * time, we would have removed all BSSs created by this driver and only
@@ -11153,15 +11154,18 @@ static bool nl80211_is_drv_shared(void *
 	/* This is the only BSS present */
 	bss = priv;
 
-	/* If only one/no link is there no one is sharing */
-	if (bss->valid_links <= 1)
-		return false;
-
 	/* More than one link means someone is still using. To check if
 	 * only 1 bit is set, power of 2 condition can be checked. */
 	if (!(bss->valid_links & (bss->valid_links - 1)))
 		return false;
 
+	/* In some scenarios, NL80211_CMD_STOP_AP from kernel can remove link
+	 * from the bss which would decrement n_links.
+	 */
+	for_each_link(bss->valid_links, i) {
+		if (i != link_id)
+			return true;
+	}
 	return true;
 }
 
--- a/src/drivers/driver.h
+++ b/src/drivers/driver.h
@@ -5394,7 +5394,7 @@ struct wpa_driver_ops {
 	/**
 	 * is_drv_shared - Check whether the driver interface is shared
 	 * @priv: Private driver interface data from init()
-	 * @bss_ctx: BSS context for %WPA_IF_AP_BSS interfaces
+	 * @link_id: Link ID
 	 *
 	 * Checks whether the driver interface is being used by other partner
 	 * BSS(s) or not. This is used to decide whether the driver interface
@@ -5402,7 +5402,7 @@ struct wpa_driver_ops {
 	 *
 	 * Returns: true if it is being used or else false.
 	 */
-	bool (*is_drv_shared)(void *priv, void *bss_ctx);
+	bool (*is_drv_shared)(void *priv, int link_id);
 
 
 	/**
